### PR TITLE
Add support for all dtypes in the tracer.

### PR DESCRIPTION
**Pull request by bgogul**

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 1]
```diff
@@ -726,6 +725,18 @@ public protocol _TensorArrayProtocolEnhanced : TensorArrayProtocol {
     where C.Element == CTensorHandle
 }
 
+
+extension _TensorArrayProtocolEnhanced {
+  func _dtypes() -> [TF_DataType] {
+    let count = _tensorHandleCount
```
Comment (mhong) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: nit: consider removing the local var `count` as it's not doing much. 

Or I'd write:
```swift
    let count = Int(_tensorHandleCount)
```

and remove the cast in the next line.

**Review by mhong — APPROVED** [review]
very nice! Look forward to bringing up data pipelines and models. :)

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 1 above)
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Use a computed property.

```swift
var _dtypes: [TF_DataType] {
  ...
}
```

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 2]
```diff
@@ -743,17 +754,15 @@ public func _graph<State : _TensorArrayProtocolEnhanced,
                            "Should not be in tracing mode already!")
 
   // Switch to tracing mode.
-  let inputValueCount = Int(state._tensorHandleCount) + Data._typeList.count
-  _RuntimeConfig.traceState = .tracing(
-    TraceContext(inputValueCount: inputValueCount))
+  let dtypes = state._dtypes() + (Data._typeList.map { $0.cDataType })
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Parentheses are redundant.

```suggestion
  let dtypes = state._dtypes() + Data._typeList.map { $0.cDataType }
```

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 3]
```diff
@@ -783,13 +792,12 @@ public func _graph<State : _TensorArrayProtocolEnhanced,
...
-      // TODO: support other dtypes.
-      Tensor<Float>(handle: TensorHandle(_owning: $0))
+      _TFCCreateTensorHandleFromC($0)
     }
     debugLog("Getting input data tensor handles.")
     let inputDataTensorHandles =  data.cTensorHandles
     inputTensors.append(contentsOf: inputDataTensorHandles.map {
-                          Tensor<Float>(handle: TensorHandle(_owning: $0))
+                          _TFCCreateTensorHandleFromC($0)
                         })
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]:
```suggestion
    })
```

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 4]
```diff
@@ -783,13 +792,12 @@ public func _graph<State : _TensorArrayProtocolEnhanced,
...
     var inputTensors = inputStateTensorHandles.map {
-      // TODO: support other dtypes.
-      Tensor<Float>(handle: TensorHandle(_owning: $0))
+      _TFCCreateTensorHandleFromC($0)
     }
     debugLog("Getting input data tensor handles.")
     let inputDataTensorHandles =  data.cTensorHandles
     inputTensors.append(contentsOf: inputDataTensorHandles.map {
-                          Tensor<Float>(handle: TensorHandle(_owning: $0))
+                          _TFCCreateTensorHandleFromC($0)
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]:
```suggestion
      _TFCCreateTensorHandleFromC($0)
```

File: test/TensorFlowRuntime/tracer.swift [hunk 5]
```diff
@@ -60,6 +60,59 @@ TracerTests.testAllBackends("Basic") {
...
+  func traceAndCheckNearlyEqual<T>(_ v: T)
+  where T: TensorFlowScalar & FloatingPoint & ExpressibleByFloatLiteral {
+    let (newState, newData) = traceAndRun(v)
+
+    expectNearlyEqualWithScalarTensor(v, newState)
+    expectNearlyEqualWithScalarTensor(v + 4, newData)
+  }
+
+  func traceAndCheckEqual<T>(_ v: T)
+  where T: TensorFlowScalar & Numeric & Comparable {
```
Comment (rxwei) [inline test/TensorFlowRuntime/tracer.swift]: Indent by 2 from `func`. Same in other places.

**Review by rxwei — APPROVED** [review]

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 1 above)
Comment (bgogul) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: Done.

Comment (bgogul) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: Done.

File: test/TensorFlowRuntime/tracer.swift (hunk 5 above)
Comment (bgogul) [reply test/TensorFlowRuntime/tracer.swift]: Done.

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 6]
```diff
@@ -726,6 +725,18 @@ public protocol _TensorArrayProtocolEnhanced : TensorArrayProtocol {
     where C.Element == CTensorHandle
 }
 
+
+extension _TensorArrayProtocolEnhanced {
+  func _dtypes: [TF_DataType] = {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: `var`, not `func`. No `=` necessary. Also, this is not used outside of this file, so make the extension be a `private extension`.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 7]
```diff
@@ -743,17 +754,15 @@ public func _graph<State : _TensorArrayProtocolEnhanced,
...
-    TraceContext(inputValueCount: inputValueCount))
+  let dtypes = state._dtypes + Data._typeList.map { $0.cDataType }
+  _RuntimeConfig.traceState = .tracing(TraceContext(dtypes: dtypes))
 
   // Handle inputs.
   let traceFn = _RuntimeConfig.traceState.context!
-  // TODO: support other dtypes.
   let inputSymbolicTensors = traceFn.symbolicInputs.map {
-    TFE_NewTensorHandleFromTFOutput($0, TF_FLOAT)!
+      TFE_NewTensorHandleFromTFOutput($0, TF_OperationOutputType($0))!
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Indent by 2 from `let`.

**Review by rxwei — APPROVED** [review]

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 6 above)
Comment (bgogul) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: > `var`, not `func`. No `=` necessary. Also, this is not used outside of this file, so make the extension be a `private extension`.

Ah, it was a typo. Thanks for catching it.

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 7 above)
Comment (bgogul) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: Done

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Comment by bgogul** [comment]
@swift-ci please test tensorflow asan

**Pull request merged**